demos/object_detection_demo/main.cpp

```diff
@@ -0,0 +1,276 @@
+        throw std::invalid_argument("Renderer: metadata is null");
+    }
+
+    auto outputImg = result.metaData->asRef<ImageMetaData>().img.clone();
+
+    if (outputImg.empty()) {
+        throw std::invalid_argument("Renderer: image provided in metadata is empty");
+    }
+
+    // Visualizing result data over source image
+
+    for (auto obj : result.objects) {
+        std::ostringstream conf;
+        conf << ":" << std::fixed << std::setprecision(3) << obj.confidence;
+        cv::putText(outputImg, obj.label + conf.str(),
+            cv::Point2f(obj.x, obj.y - 5), cv::FONT_HERSHEY_COMPLEX_SMALL, 1,
+            cv::Scalar(0, 0, 255));
+        cv::rectangle(outputImg, obj, cv::Scalar(0, 0, 255));
+    }
+
+    return outputImg;
+}
+
+
+int main(int argc, char *argv[]) {
+    try {
+        PerformanceMetrics metrics;
+
+        slog::info << "InferenceEngine: " << printable(*InferenceEngine::GetInferenceEngineVersion()) << slog::endl;
+
+        // ------------------------------ Parsing and validation of input args ---------------------------------
+        if (!ParseAndCheckCommandLine(argc, argv)) {
+            return 0;
+        }
+
+        //------------------------------- Preparing Input ------------------------------------------------------
+        slog::info << "Reading input" << slog::endl;
+        auto cap = openImagesCapture(FLAGS_i, FLAGS_loop);
+        cv::Mat curr_frame;
+
+        //------------------------------ Running Detection routines ----------------------------------------------
+        std::vector<std::string> labels;
+        if (!FLAGS_labels.empty())
+            labels = DetectionModel::loadLabels(FLAGS_labels);
+
+        std::unique_ptr<ModelBase> model;
+        if (FLAGS_at=="ssd") {
+            model.reset(new ModelSSD(FLAGS_m, (float)FLAGS_t, FLAGS_auto_resize, labels));
+        }
+        else if (FLAGS_at=="yolo") {
+            model.reset(new ModelYolo3(FLAGS_m,(float)FLAGS_t, FLAGS_auto_resize, (float)FLAGS_iou_t, labels));
+        }
+        else {
+            slog::err << "No model type or invalid model type (-at) provided: " + FLAGS_at << slog::endl;
+            return -1;
+        }
+
+        InferenceEngine::Core core;
+        AsyncPipeline pipeline(std::move(model),
+            ConfigFactory::getUserConfig(FLAGS_d, FLAGS_l, FLAGS_c, FLAGS_pc, FLAGS_nireq, FLAGS_nstreams, FLAGS_nthreads),
+            core);
+        Presenter presenter;
+
+        bool keepRunning = true;
+        while (keepRunning){
+            int64_t frameNum = 0;
+            if (pipeline.isReadyToProcess()) {
+                //--- Capturing frame. If previous frame hasn't been inferred yet, reuse it instead of capturing new one
+                auto startTime = std::chrono::steady_clock::now();
+                curr_frame = cap->read();
+                if (curr_frame.empty()) {
+                    if (!frameNum) {
+                        throw std::logic_error("Can't read an image from the input");
```

Comment:
For some reason I can't reply in https://github.com/openvinotoolkit/open_model_zoo/pull/1639#discussion_r526802743 So I start a new conversation :)

> cap->read may return empty Mat in case of error.

You can't actually say if the frame is empty because of error or because of stream end. Any way the problem is that you have wrong order. You do the following:
1. init `frameNum` with 0
2. check `!frameNum` which is equivalent to `!0`
3. overwrite `frameNum`
4. go to 1.

Reply:
OK, I see now. Fixed.